include/nbl/builtin/hlsl/cmath.hlsl

```diff
@@ -0,0 +1,258 @@
+// Copyright (C) 2022 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_CMATH_INCLUDED_
+#define _NBL_BUILTIN_HLSL_CMATH_INCLUDED_
+
+#include <nbl/builtin/hlsl/limits.hlsl>
+
+// C++ headers
+#ifndef __HLSL_VERSION
+#include <cmath>
+#endif
+
+namespace nbl
+{
+namespace hlsl
+{
+
+
+#ifdef __cplusplus
+#define INTRINSIC_NAMESPACE(x) std::x
+#else
+#define INTRINSIC_NAMESPACE(x) x
+#endif
+
+#define NBL_ALIAS_BINARY_FUNCTION2(name,impl) template<class T> T name(T x, T y) { return INTRINSIC_NAMESPACE(impl)(x, y); }
+#define NBL_ALIAS_BINARY_FUNCTION(fn) NBL_ALIAS_BINARY_FUNCTION2(fn,fn)
+#define NBL_ALIAS_UNARY_FUNCTION2(name,impl)  template<class T> T name(T x) { return INTRINSIC_NAMESPACE(impl)(x); }
+#define NBL_ALIAS_UNARY_FUNCTION(fn)  NBL_ALIAS_UNARY_FUNCTION2(fn,fn)
```

Comment:
life is never that easy, since we're bridging the gap between HLSL and STL and having a unified namespace for lang/target agnostic code

#### C++

In HLSL all those functions must be able to work on `vector`, so C++ needs a per-component "forwarder" to do the function on each component of a vector when `is_vector` is true.

Also C++ has templated functions in `tgmath` and I'd prefer you to use that rather than rely on overload resolution to hit the right one.

### HLSL

Check if the intrinsics don't have templated versions (cause apparently `ByteAddressBuffer` does and they're not in the documentation XD).

### Something to think about

#### How will this all with `complex`

Does `tgmath` need to include `complex` ? Should we forward declare the specializations?

Remember that `complex` will be templated not only on `float_t` but `fvec_t` and e.g. `complex<float32_t3>` is not a builtin type so will need to be forwarded component by component even in HLSL.

#### After OpenEXR upgrade: Use IlmMath's versions of all these functions if it has them

HLSL doesn't have all these functions for `half` types (only GLSL on AMD used to).

For now its okay to not support any of this for HLSL and C++.